File: Session13/320. Generalized Abbreviation.py
class Solution:
    def generateAbbreviations(self, word: str) -> List[str]:
        n=len(word)
        ans=[]
        temp=[]

        def helper(pos,k):
            nonlocal temp
            l=len(temp)
            if pos>=n:
                if k!=0:
                    temp.append(str(k))
                ans.append("".join(temp))
                temp=temp[0:l]
                return
            helper(pos+1,k+1)
            if k!=0:
                temp.append(str(k))
            temp.append(word[pos])
            helper(pos+1,0)
            temp=temp[0:l]
        helper(0,0)
        return ans

File: Session13/320. Generalized Abbreviation.java
class Solution {
    List<String> ans=new ArrayList<>();
    public List<String> generateAbbreviations(String word) {
       backtrack(0,new StringBuilder(),word,0);
        return ans;
    }
    
    public void backtrack(int pos,StringBuilder curr,String word,int bef_count)
    {
        if(pos==word.length())
        {
            if(bef_count!=0)
            {
                curr.append(String.valueOf(bef_count));
            }
            ans.add(curr.toString());
            return;
        }
        // Decision 1 - keep counting
        int len=curr.length();
        backtrack(pos+1,curr,word,bef_count+1);
        curr.setLength(len);
        
        // Decision 2 - Break counting, keep me as a character, start new counting from next position
        if(bef_count!=0)
        {
            curr.append(String.valueOf(bef_count));
        }
        curr.append(word.charAt(pos));
        backtrack(pos+1,curr,word,0);
        curr.setLength(len);
    }
}